base/src/H264DecoderNvCodecHelper.cpp

```diff
@@ -0,0 +1,754 @@
+			}
+		}
+		T* pv = puv + (nPitch / 2) * (nHeight / 2);
+		for (int y = 0; y < nHeight / 2; y++) {
+			for (int x = 0; x < nWidth / 2; x++) {
+				puv[y * nPitch + x * 2] = pQuad[y * nWidth / 2 + x];
+				puv[y * nPitch + x * 2 + 1] = pv[y * nPitch / 2 + x];
+			}
+		}
+	}
+	void UVInterleavedToPlanar(T* pFrame, int nPitch = 0) {
+		if (nPitch == 0) {
+			nPitch = nWidth;
+		}
+		T* puv = pFrame + nPitch * nHeight,
+			* pu = puv,
+			* pv = puv + nPitch * nHeight / 4;
+		for (int y = 0; y < nHeight / 2; y++) {
+			for (int x = 0; x < nWidth / 2; x++) {
+				pu[y * nPitch / 2 + x] = puv[y * nPitch + x * 2];
+				pQuad[y * nWidth / 2 + x] = puv[y * nPitch + x * 2 + 1];
+			}
+		}
+		if (nPitch == nWidth) {
+			memcpy(pv, pQuad, nWidth * nHeight / 4 * sizeof(T));
+		}
+		else {
+			for (int i = 0; i < nHeight / 2; i++) {
+				memcpy(pv + nPitch / 2 * i, pQuad + nWidth / 2 * i, nWidth / 2 * sizeof(T));
+			}
+		}
+	}
+
+private:
+	T* pQuad;
+	int nWidth, nHeight;
+};
+
+H264DecoderNvCodecHelper::H264DecoderNvCodecHelper(int mWidth, int mHeight)
+{
+	CUcontext cuContext;
+	bool bUseDeviceFrame = false;
+	cudaVideoCodec eCodec = cudaVideoCodec_H264;;
+	std::mutex* pMutex = NULL; 
+	bool bLowLatency = false; 
+	bool bDeviceFramePitched= false; int maxWidth; int maxHeight;
+	ck(cuInit(0));
+	int nGpu = 0;
+	int iGpu = 0;
+	ck(cuDeviceGetCount(&nGpu));
+	if (iGpu < 0 || iGpu >= nGpu) {
+		std::cout << "GPU ordinal out of range. Should be within [" << 0 << ", " << nGpu - 1 << "]" << std::endl;
+		return;
+	}
+	CUdevice cuDevice = 0;
+
+	ck(cuDeviceGet(&cuDevice, iGpu));
+	char szDeviceName[80];
+	ck(cuCtxCreate(&cuContext, 0, cuDevice));
+	helper.reset(new NvDecoder(cuContext, mWidth, mHeight, bUseDeviceFrame, eCodec, pMutex));
+}
+
+bool H264DecoderNvCodecHelper::init(std::function<void(frame_sp&)> _send, std::function<frame_sp()> _makeFrame)
+{
+	makeFrame = _makeFrame;
+	send = _send;
+	return false;
+}
+
+void H264DecoderNvCodecHelper::ConvertToPlanar(uint8_t* pHostFrame, int nWidth, int nHeight, int nBitDepth) {
+	if (nBitDepth == 8) {
+		// nv12->iyuv
+		YuvConverter<uint8_t> converter8(nWidth, nHeight);
+		converter8.UVInterleavedToPlanar(pHostFrame);
+	}
+	else {
+		// p016->yuv420p16
+		YuvConverter<uint16_t> converter16(nWidth, nHeight);
+		converter16.UVInterleavedToPlanar((uint16_t*)pHostFrame);
+	}
+}
+
+bool H264DecoderNvCodecHelper::process(frame_sp& frame)
+{
+	uint8_t* inputBuffer = NULL;
+	int inputBufferSize = 0;
+	frame_sp outputFrame = makeFrame();
+	uint8_t** outBuffer = reinterpret_cast<uint8_t**>(outputFrame->data());
+	
+	inputBuffer = static_cast<uint8_t*>(frame->data());
+	inputBufferSize = frame->size();
+	
+	int nFrameReturned = 0, nFrame = 0;
+	bool bOutPlanar = true;
+
+	helper->Decode(inputBuffer, inputBufferSize, &outBuffer, &nFrameReturned);
+
+	for (int i = 0; i < nFrameReturned; i++)
+	{
+		ConvertToPlanar(outBuffer[i], helper->GetWidth(), helper->GetHeight(), helper->GetBitDepth());
```

Comment:
Line 735 creates a single outFrame so how can we use it in a loop of 'nFrameReturned" ?
have you ever tried using it in a scenario where nFrameReturned >1 ?

Reply:
Yes i've tried the scenario where nFameReturned > 1. Yes , only single OutFrame is created but then i'm over writing on the same outFrame->data() and send() the frame to next module whenever nFrameReturned > 1.